ERCC refactoring

mbrandenburger opened the pull request:
Signed-off-by: bur <bur@zurich.ibm.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our code of conduct and contributor guidelines: 
     https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/CONTRIBUTING.md
     https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/CODE_OF_CONDUCT.md
   In particular pay attention to the git workflows
      https://docs.google.com/document/d/1sR7YV3pSYN3NEFiW-2fUqtpsJeJrpC0EWUVtEm0Blcg/edit#heading=h.kwcug3pkefak
2. Fill out below sections.
3. Label the PR with the label of any component this PR touches.
4. ALso don't forget to sign your comments before submitting. 
   Github will complain if there is no DCO but it's easier if we don't have to hunt you down to fix that :-)

-->

**What this PR does / why we need it**:
- implements RegiterEnclave
- implemented as a chaincode as a service

**Which issue(s) this PR fixes**:
<!--
  list existing bug, feature and/or work-item which this PR addresses.
  You might also consider creating an issue first for the PR.
-->
Addresses ERCC related work items in #404
**Special notes for your reviewer**:

**Does this PR introduce a user-facing changes and/or breaks backward compatability?**:
<!--
  If no, you can delete this section
  If yes, describe what changes and/or what breaks ..
-->
```

g2flyer commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,255 @@
+	chaincodeId := attestedData.CcParams.ChaincodeId
+	enclaveId := utils.GetEnclaveId(&attestedData)
+
+	key, err := ctx.GetStub().CreateCompositeKey("namespaces/credentials", []string{chaincodeId, enclaveId})
```
Wouldn't you have to do here also some check whether there are already registered enclaves for that chaincode and, if so, verify that they are consistent?

mbrandenburger replied:
added todos

g2flyer commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,255 @@
+	// TODO add more checks
+	// channel_hash should correspond to peers view of channel id (POST-MVP)
+	// TLCC_MRENCLAVE matches the version baked into ERCC (POST-MVP)
+	// check org-enclave binding (POST-MVP)
```
hmm, didn't we just check that with above HostParams checks?
```suggestion
```

g2flyer commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,255 @@
+	// channel_hash should correspond to peers view of channel id (POST-MVP)
+	// TLCC_MRENCLAVE matches the version baked into ERCC (POST-MVP)
+	// check org-enclave binding (POST-MVP)
+	// deployment validation (POST-MVP)
+	// validate FPC deployment (restriction) policy
```
what's the difference between the two?

mbrandenburger replied:
same

g2flyer commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,255 @@
+}
+
+// returns the chaincode encryption key for a given chaincode id
+func (rs *Contract) QueryChaincodeEncryptionKey(ctx contractapi.TransactionContextInterface, chaincodeId string) ([]byte, error) {
+	//input chaincodeId string
+
+	//return chaincode_ek []byte
+	return nil, nil
```
Hmm, some TODO is missing here (or even better, the corresponding implementation: should be straightforward?

mbrandenburger replied:
I added a TODO.

g2flyer replied:
Can't we just implement it? should be pretty similar to the other queries?

mbrandenburger replied:
I can be your PR once this here is merged!

g2flyer replied:
well, we can do it in a separate PR; just seemed random why we do all the others but not these and add the overhead of a separate PR? But i guess as long as we have the TODO my main concern is addressed

g2flyer reviewed (commented):
Some initial comments.  I haven't looked at the testing stuff, you have to give a bit more context on that (see corresponding comment).

mbrandenburger commented on ercc/connection.json:
```diff
@@ -0,0 +1,5 @@
+{
```
I agree, see d14b41c, it is already moved somewhere.

mbrandenburger commented on ercc/attestation/interface.go:
```diff
@@ -0,0 +1,5 @@
+package attestation
+
+type VerifierInterface interface {
+	VerifyEvidence(evidenceBytes, expectedStatementBytes []byte, expectedMrEnclave string) error
+}
```
right, I also think that we will only have a single implementation of this interface that contains the cgo part calling the real attestation api. However, having this interface is just here to autogenerate mock stubs and use them in testing.

mbrandenburger commented on ercc/chaincode.env:
```diff
@@ -0,0 +1,10 @@
+# used to get the ID after install if required
+CHAINCODE_ID=ercc_1.0:26135f3588187d357d35b2e3b6e1c3b2805860815c10aa37b1e90a3b44ba47d2
+
+FABRIC_LOGGING_SPEC=DEBUG
```
ack! moved

mbrandenburger commented on ercc/registry/mocks/chaincodestub.go:
```diff
@@ -0,0 +1,2878 @@
+// Code generated by counterfeiter. DO NOT EDIT.
```
yes correct, theres files are generated using `go generate ./...` based on the interfaces specified in `registry_test.go` using [counterfeiter](https://github.com/maxbrunsfeld/counterfeiter). I added it also to our tools.go, so it should be already there.

I think for the test mocks here we can follow the fabric approach and check these generated files in. However, we can come back on that later.

For documentation I only used https://github.com/maxbrunsfeld/counterfeiter and looking around in the fabric repo, for instance [here](https://github.com/hyperledger/fabric/tree/master/core/chaincode).

g2flyer replied:
what is the reasoning to check in the mocks?! and what makes sure that they are kept in sync with `registry_test.go`. Honestly i don't really like that. At the very least we should have a make target which creates them and they shouldn't appear out of thin air without documentation ...

mbrandenburger commented on internal/utils/endorsement.go:
```diff
@@ -3,14 +3,22 @@ package utils
 )
 
-func NewPolicyEvaluator() *PolicyEvaluator {
+type PolicyEvaluatorInterface interface {
```
we have an interface here to add mocks so in `registry_test.go` we can better control the behaviour of the policy evaluator. I think this is also the normal way to go in fabric. Once you implement a component that is composed of other components, you better use interfaces so you can concentrate on the implementation and testing of the component you are developing.

ok, the main reason for the change here is that the previous implementation of `EvaluateIdentity` relied on a identityDeserializer provided by default msp provider (not sure if this is the right name?!?! anyway). In particular, the old version called the following code to get the deserializer.

```go
manager := mgmt.NewDeserializersManager(factory.GetDefault())
eturn &PolicyEvaluator{
		IdentityDeserializer: manager.GetLocalDeserializer(),
``` 

However, using this code a chaincode environment we get a deserializer that cannot parse the identities simply because the chaincode environment has no access to the msp information that a peer has. 

Therefore, this change here. What we do now is essentially, unmarshal the policy and the serialized identity and check if the msp id of is part of the policy.

mbrandenburger replied:
will add a few comments

mbrandenburger replied:
Not sure what you mean with different implementation of this evaluator?

g2flyer replied:
Actually, i think i know where the disconnect comes. Originally before lightweight the chaincode endorsement policy was really policy which decided on validation rules for attestations, i.e., what in our proposal we call `enclave endorsement` policies.  However, now we have this `enclave endorsement policy` separate from the `validation endorsement policy` and the endorsement policy of the chaincode is the latter whereas the former is for MVP `designated enclave` policy implicitly defined  (and for post-mvp specified in some yet to be determined fashion).  What you reference/verify here is hence against the `validation endorsement policy`.  However, what we do have to verify in errc is the `enclave endorsement policy`.  Hence my reference to above that i was speculating that we might have (eventually) multiple implementations depending on what enclave endorsement policy we support. Not that i think this is really necessary and i now also understand why you have an interface even if there is only one implementation, the current implemention imho is doing the wrong stuff (at least as used by ercc)

mbrandenburger commented on internal/utils/chaincode.go:
```diff
@@ -18,14 +18,14 @@ import (
 const MrEnclaveLength = 32
 
-func GetChaincodeDefinition(chaincodeId string, stub shim.ChaincodeStubInterface) (*lifecycle.QueryChaincodeDefinitionResult, error) {
+func GetChaincodeDefinition(chaincodeId, channelId string, stub shim.ChaincodeStubInterface) (*lifecycle.QueryChaincodeDefinitionResult, error) {
```
mh, yeah I think this change is not really necessary. I guess some relict some playing with the code :)

mbrandenburger commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,255 @@
+// - query `getEnclaveId` at all the peers discovered
+// - query `queryListEnclaveCredentials` with all received enclave_ids
+// this gives you the endpoints and credentials including enclave_vk, and chaincode_ek
+func (rs *Contract) QueryListEnclaveCredentials(ctx contractapi.TransactionContextInterface, chaincodeId string) ([][]byte, error) {
```
so by default the contract API only support a bunch of input and output primitive data types. Our credential data type is not supported. There is a way to provide custom serializer to make this possible but I haven't yet looked into it. 

Also, maybe we also don't care in this function because getState(ByPar...) returns a serialized Credentials object. So if this function would return an array of []Credentials. we would need to deserialize them after getState and returns. However, the custom serializer would serialize them again. Let me add more comments to the code.

g2flyer replied:
ok, completely missed when looking at it that these parameters come from the request, not one of our other go functions, and hence are the encoded version of these objects (i.e., serialized protobuf and as discussed in this morning calls also base64 encoded (on top) so on the client side we have strings which can be handled by peer cli script.
Clarifying that in documentation though would help. => all add a corresponding proposal for a change ...

mbrandenburger commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,255 @@
+	return allCredentials, nil
+}
+
+func (rs *Contract) QueryEnclaveCredentials(ctx contractapi.TransactionContextInterface, chaincodeId, enclaveId string) ([]byte, error) {
```
see comment above

mbrandenburger commented on ercc/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+FROM hyperledger/fabric-private-chaincode-base-rt:latest
```
yes, we should move the Dockerfile somewhere else. I would do that in a follow up PR once it's clear where

g2flyer replied:
as mentioned, it should be part of the sample net stuff, i.e., in that pr. And regarding location i would suggest we put all of the sample net into 'util/sample-net' or alike as this is also were we have docker and docker-compose, with the latter presumably eventually replaced by the sample-network stuff ...

g2flyer commented on ercc/Makefile:
```diff
@@ -6,32 +6,28 @@
-docker-clean:
-	-if [ ! -z "$(DOCKER_IMAGE)" ]; then docker rmi -f $(DOCKER_IMAGE); fi
-	-if [ ! -z "$(DOCKER_IMAGE)" ]; then echo "\033[0;33mWARNING: ercc not deleted (peer still up?)\033[0m"; fi
+run:
```
you moved only the docker stuff. I was referring to also all the packaging stuff which is also sample-net-test specific ...

g2flyer replied:
the docker part though imho is also part of sample-net-test part.  But obviously that has to be handled together with whereever the Docker file is

g2flyer commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,269 @@
+		return err
+	}
+
+	// TODO check if this enclave is already registered
```
what are we doing if already registered? fail?

mbrandenburger replied:
good question, let's figure out later.

g2flyer commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,269 @@
+
+	// TODO check if this enclave is already registered
+
+	// TODO perform the MVP check that only a single enclave is registered per chaincode
```
```suggestion
	// TODO perform the (enclave) endorsement policy specific tests:
	// - MVP (designated chaincode): verify no other enclave (of other peers) is registered or fail
	// - Post-MVP: check consistency with potentially existing enclaves
```

g2flyer commented on ercc/registry/registry_test.go:
```diff
@@ -0,0 +1,339 @@
+	"github.com/stretchr/testify/require"
+)
+
+//go:generate counterfeiter -o mocks/transaction.go -fake-name TransactionContext . transactionContext
```
was there a strong reason change the outpudle directory and fake-name?  I actually like the default name and distinguishes from the normal mocks (see my other comment that i got confused by the name "mock" because this mock doesn't match the other mocks which are usually self-contained functional replacements for a module rather than the kindof  "remote-controlled" ones here.   That said, no strong reason but it confused me and using defaults even makes these comments more consice?

mbrandenburger replied:
first reason, I got inspired by some package in fabric using counterfeiter with this naming convention. And by looking at the output of the default setting, I like this non-standard format more. I think we maybe rename the `mocks` folder to `fakes` to do something against you mental picture of mocks and "remote-controlled" fakes. :)

CHanging also the interface names to default with FakeEverything naming I don't like so much as the fake package already notes that this is a "mock" or "fake" implementation.

g2flyer commented on ercc/registry/registry_test.go:
```diff
@@ -0,0 +1,339 @@
+)
+
+//go:generate counterfeiter -o mocks/transaction.go -fake-name TransactionContext . transactionContext
+type transactionContext interface {
+	contractapi.TransactionContextInterface
+}
```
Wouldn't this in principle be unecessary if you use the third-party variant of generation where the argument is <package-path>.<Interface>. (Not that i'm asking for change, just trying to understand ....)

mbrandenburger replied:
maybe. cannot answer this without trying out.

g2flyer commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,271 @@
+//
+// Note that this implementation returns a set of serialized credentials in order to send it to the receiver.
+// That is, the receiver needs to deserialize the return value into []Credentials
+// TODO look into custom serializer to make this call more aligned with the FPC API in `interfaces.md`
```
well, we didn't really define the encoding in `interfaces.md` (there is only a "maybe" which is anyway different from what we talk now).  From discussion this morning it seems we have agreed on base64-encoding of (standard binary) protobuf encoding of Objects?

g2flyer commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,255 @@
+// This method is used during the key generation and key distribution protocol. In particular, during key generation,
+// this call sets the chaincode_ek for a chaincode if no chaincode_ek is set yet.
+func (rs *Contract) RegisterCCKeys(ctx contractapi.TransactionContextInterface, ccKeyRegistrationMessageBase64 string) error {
+	//input msg CCKeyRegistrationMessage
```
It seems you approved the comment but didn't add the TODO?

g2flyer commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,271 @@
+// - query `queryListEnclaveCredentials` with all received enclave_ids
+// this gives you the endpoints and credentials including enclave_vk, and chaincode_ek
+//
+// Note that this implementation returns a set of serialized credentials in order to send it to the receiver.
```
Might as well make the encoding explicit?
```suggestion
// Note that this implementation returns a set of (base64-encoded) protobuf-serialized `Credential` objects in order to send it to the receiver.
```

g2flyer commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,271 @@
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package registry
```
```suggestion
// this package defines the client-facing interface of ERCC as defined in the ERCC Interface section in [specifications](https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/docs/design/fabric-v2%2B/interfaces.md)
package registry
```

g2flyer commented on ercc/registry/registry.go:
```diff
@@ -0,0 +1,271 @@
+// registers a CCKeyRegistration message that confirms that an enclave is provisioned with the chaincode encryption key.
+// This method is used during the key generation and key distribution protocol. In particular, during key generation,
+// this call sets the chaincode_ek for a chaincode if no chaincode_ek is set yet.
+func (rs *Contract) RegisterCCKeys(ctx contractapi.TransactionContextInterface, ccKeyRegistrationMessageBase64 string) error {
+	//input msg CCKeyRegistrationMessage
```
```suggestion
func (rs *Contract) RegisterCCKeys(ctx contractapi.TransactionContextInterface, ccKeyRegistrationMessageBase64 string) error {
         // TODO: Implement me for MVP
         
	//input msg CCKeyRegistrationMessage
```

g2flyer reviewed (changes requested):
Overall great step forward for ERCC and probably (actually, almost certainly?) the most comprehensive test we have anywhere in FPC :-)

However, i think there is an issue with the policy evaluation where the old model of (chaincode) endorsement policies doesn't really represent the (enclave) endorsement policy we had in our spec and hence the implementation seems to do somewhat the wrong checks. See also some more text in the discussion on `endorsement.go`. That said, probably, best discussed in some interactive way instead of going through github?

We also should definitely add the Makefile target for the counterfeit generation.  I actually tried to add it as a proposal but then run into the problem i simply cannot build FPC right now (and hence not test any make target as `go generate` also failed).  I understand that integration tests and alike do not work but shouldn't `make build` work?  Do i do something wrong or did you maybe forget to check something in?
```
...
/hyperledger/fabric-contract-api-go v1.1.1
# github.com/hyperledger-labs/fabric-private-chaincode/internal/utils
../internal/utils/utils.go:65:33: undefined: protos
../internal/utils/utils.go:66:7: undefined: sha256
../internal/utils/utils.go:67:9: undefined: hex
Makefile:12: recipe for target 'ercc' failed
```

mbrandenburger commented:
Thanks @g2flyer. I updated the `fakes` generation, including renaming and add a build target for go generate. Fixed also the include issue in `utils.go`.

Let's have a quick sync offline to resolve the identified issue regarding EP validation.

g2flyer commented:
FYI: tried again to build after your update and i'm still running in issues with `make build`. 
- it fails in fabric during peer build in validators. I've solved it by a close-to-minimal fix of moving `fabric/patches/3_0001-Add-FPC-validator-to-library.patch` to `fabric/patches/3_0001-Add-FPC-validator-to-library.patch` after which fabric build. (Alternatively we could of course for now also just remove `fabric` from the subdir list in the top-level makefile) 
- `protos` has to be added to `FPC_SDK_DEP_DIRS` in the top-level makefile is needed or cc-builder fails to build
With above two changes, `make build` does work successfully from `$FPC_PATH`

For `make test` i would expected it to fail in `demo` (which we probably should just comment out for now in Makefile) and probably in `integration-test` (maybe also comment out for now?) but it already failed very quickly when trying to build `internal`:
```
....
ok      github.com/hyperledger-labs/fabric-private-chaincode/internal/testing/executor  (cached)
?       github.com/hyperledger-labs/fabric-private-chaincode/internal/testing/executor/worker   [no test files]
# github.com/hyperledger-labs/fabric-private-chaincode/internal/utils_test [github.com/hyperledger-labs/fabric-private-chaincode/internal/utils.test]
utils/chaincode_test.go:54:44: too many arguments in call to utils.GetChaincodeDefinition
        have (string, string, *shimtest.MockStub)
        want (string, shim.ChaincodeStubInterface)
utils/chaincode_test.go:62:44: too many arguments in call to utils.GetChaincodeDefinition
        have (string, string, *shimtest.MockStub)
        want (string, shim.ChaincodeStubInterface)
utils/chaincode_test.go:76:44: too many arguments in call to utils.GetChaincodeDefinition
        have (string, string, *shimtest.MockStub)
        want (string, shim.ChaincodeStubInterface)
FAIL    github.com/hyperledger-labs/fabric-private-chaincode/internal/utils [build failed]
FAIL
Makefile:9: recipe for target 'test' failed
make[1]: *** [test] Error 2
make[1]: Leaving directory '/home/msteiner/src/projects/intel/PoET-et-al/fabric/go-trusted-chain-code.v2.0/src/github.com/hyperledger-labs/fabric-private-chaincode/internal'
Makefile:22: recipe for target 'test' failed
```


BTW: just noticed that you called the branch `ercc-refactor`. Just to make sure, this is not an ercc-specific branch but is the intended 'refactor-to-new-ecc-ercc+lightweight' branch we talked earlier?

mbrandenburger commented:
Regarding building fabric peer this is a good observation! We can disable it for the moment and also add proto.

Regarding the issue in utils, this should be resolved.

bvavala commented on internal/utils/endorsement.go:
```diff
@@ -3,40 +3,89 @@ package utils
+	return &PolicyEvaluator{}
+}
+
+// EvaluateIdentity checks that the provided identity is a valid endorser as specified in the endorsement policy
```
My understanding is that this part is to check whether the enclave is a valid endorser.

That said, do we need this whole part at all?

We definitely _need_ that in an FPC/Fabric integration where the enclave acts as a _Fabric_-endorser and signs the response.
However, in the new architecture, this is not the case anymore. The enclave acts as an _FPC_-endorser and its signature is checked by the FPC (_not Fabric_) validation chaincode. Most importantly, FPC uses ERCC to control which enclaves can FPC-endorse a transaction and, therefore, pass the validation check. Hence, it is not necessary to check whether the Fabric validation logic can recognize an enclave's signature through MSP.

mbrandenburger replied:
yes, all correct. We can keep this code for FULL FPC ercc implementation. Note that the policy evaluated got invalided with c4a0ace during review session with Michael :D

bvavala reviewed (commented):
Some comments and a question about endorser validation.

mbrandenburger commented on ercc/Makefile:
```diff
@@ -6,32 +6,23 @@
-docker-clean:
-	-if [ ! -z "$(DOCKER_IMAGE)" ]; then docker rmi -f $(DOCKER_IMAGE); fi
-	-if [ ! -z "$(DOCKER_IMAGE)" ]; then echo "\033[0;33mWARNING: ercc not deleted (peer still up?)\033[0m"; fi
+docker:
```
oh this docker-clean cleans ercc images produced by a peer. Since we use either chaincode as a service or chaincode started using external builder this special clean command is not needed anymore. But I think a docker-clean is still helpful here. Let me bring back a docker clean.

mbrandenburger commented:
Squashed and ready for merge.
